Add  events to indicate start and end of skaffold dev iterations

tejal29 opened the pull request:
<!-- Thanks you for your contribution! -->

<!-- Include if applicable: -->
/cc @quoctruong
**Fixes**: #4036
**Related**: _Relevant tracking issues, for context_
**Merge before/after**: _Dependent or prerequisite PRs_

**Description**
<!-- Describe your changes here. The more detail, the easier the review! -->
In this PR, 
**User facing changes (remove if N/A)**
No user facing changes.
Tools consuming Event API will see following new events in the sequence.
```
tejaldesai@tejaldesai-macbookpro2 microservices (add_skaffold_end_event)curl localhost:50052/v1/events
{"result":{"timestamp":"2020-04-28T06:16:30.579021Z","event":{"metaEvent":{"entry":"Starting Skaffold: \u0026{Version:v1.8.0-67-ge5e6a38f3-dirty ConfigVersion:skaffold/v2beta3 GitVersion: GitCommit:e5e6a38f3abd6273ab14b6e7a71c41ec58b8b363 GitTreeState:dirty BuildDate:2020-04-27T23:16:05Z GoVersion:go1.14.1 Compiler:gc Platform:darwin/amd64}","metadata":{"build":{"numberOfArtifacts":2,"builders":[{"type":"DOCKER","count":2}],"type":"LOCAL"},"deploy":{"deployers":[{"type":"KUBECTL","count":1}],"cluster":"GKE"}}}}}}

{"result":{"timestamp":"2020-04-28T06:16:30.579067Z",
   "event":{"devLoopEvent":{"status":"In Progress"}},
  "entry":"Dev Iteration 0 in progress"}}
....

{"result":{"timestamp":"2020-04-28T06:16:35.192693Z",
   "event":{"devLoopEvent":{"status":"Succeeded"}},
   "entry":"Dev Iteration 0 successful"}}

{"result":{"timestamp":"2020-04-28T06:16:59.094247Z",
   "event":{"devLoopEvent":{"iteration":1, "changeType":"BUILD", "status":"In Progress"}},
   "entry":"Dev Iteration 1 in progress"}}

### COMPILE FAILURE - CAUSING A BUILD FAILURE
{"result":{"timestamp":"2020-04-28T06:17:03.748666Z",
   "event":{"buildEvent":{"artifact":"leeroy-app",
    "status":"Failed",
    "err":"unable to stream build output: The command '/bin/sh -c go build -o /app .' returned a 
         non-zero code: 2",
    "errCode":"ErrorCode_UNKNOWN"}},
   "entry":"Build failed for artifact leeroy-app"}}

{"result":{"timestamp":"2020-04-28T06:17:03.748736Z","
   event":{"devLoopEvent":{"iteration":1,"status":"Failed",
   "err":"couldn't build \"leeroy-app\": unable to stream build output: The command '/bin/sh -c go 
                build -o /app .' returned a non-zero code: 2",
   "errCode":"ErrorCode_UNKNOWN"}},
  "entry":"Dev Iteration 1 failed with error code ErrorCode_UNKNOWN"}}


{"result":{"timestamp":"2020-04-28T06:17:51.727711Z",
   "event":{"devLoopEvent":{"iteration":2 "changeType":"BUILD" "status":"In Progress"}},
    "entry":"Dev Iteration 2 in progress"}}
...
{"result":{"timestamp":"2020-04-28T06:18:12.363222Z",
   "event":{"devLoopEvent":{"status":"Succeeded"}},
    "entry":"Dev Iteration 2 successful"}}

{"result":{"timestamp":"2020-04-30T23:49:16.348194Z",
  "event":{
   "endEvent":{
      "errCode":1,"
        loops":[
          {"errCode":"SUCCESS"},   // 0 value for interation and changeType Initial Loop is ommitted in output
         {"iteration":1,  "changeType":"BUILD",  "errCode":"ErrorCode_UNKNOWN"}, 
         {"iteration":2,"changeType":"BUILD","errCode":"SUCCESS"}
        ]}},
     "entry":"Skaffold end event with 3 total dev loops"}}


curl: (18) transfer closed with outstanding read data remaining

```

**Follow-up Work (remove if N/A)**
<!-- Mention any related follow up work to this PR. -->
- [ ]Testing with IDES.


<!--
Please be sure your PR includes unit tests - we don't merge code that brings down test coverage! 
Integration tests are sometimes an appropriate substitute. 
-->

briandealwis commented on proto/skaffold.proto:
```diff
@@ -344,11 +353,32 @@ enum ErrorCode {
+    // FILE_SYNC Failures
+
+    // File Sync Initialize failure
+    FILE_SYNC_INIT_ERROR = 601;
```
Do we need the _ERROR suffix?  Isn't it implicit?

tejal29 replied:
dropped

tejal29 replied:
Change this to FILE_SYNC_INIT_FAIL

briandealwis commented on pkg/skaffold/runner/dev.go:
```diff
@@ -60,7 +63,8 @@ func (r *SkaffoldRunner) doDev(ctx context.Context, out io.Writer) error {
 	defer r.monitor.Reset()
 	defer r.listener.LogWatchToUser(out)
-
+	event.DevLoopInProgress(iteration)
```
Should these not follow the var-for-testing approach used for fileSync*?

tejal29 replied:
That is a good question.  I also thought, i would have to add test variables. 
But in the same func below, the `event.ResetStateOnDeploy` isn't mocked out in tests. The unit tests pass so i did not add variables for these.

briandealwis commented on pkg/skaffold/errors/errors.go:
```diff
@@ -21,14 +21,15 @@ import (
-func ErrorCodeFromError(_ error, _ phase) proto.ErrorCode {
-	return proto.ErrorCode_COULD_NOT_DETERMINE
+func ErrorCodeFromError(_ error, _ Phase) proto.ErrorCode {
+	return proto.ErrorCode_ErrorCode_UNKNOWN
```
`COULD_NOT_DETERMINE` seems more apt here?

It feels like we should have an error type that carries the ErrorCode inside, rather than returning an error-code and an error.  Then we could get rid of `STATUS_CHECK_NO_ERROR`.

tejal29 replied:
I think it would make sense to add  a new proto `ErrorDef` with those 2 fields. However, we won't be able to change that for other events `BuildEvent`, `DeployEvent` , `StatusCheckEvent` etc since IDEs are already consuming it. 

Re: `STATUS_CHECK_NO_ERR`, it was something we discussed for regarding following HTTP error codes for success.  It would also be easy for us to collect metrics if we define a success code vs "0" since the filed could be 0 is the default value for int type enum

briandealwis replied:
Looking back through this, I'm confused by the intent of this function.  It's a placeholder?  Can we hook in the pattern-matching aspects of #4045 here?

Or we we could our own `error` that carries an ErrorCode?  Like the [`os.Err*`](https://golang.org/pkg/os/#pkg-variables) variables and `os.IsExist()`.
<details>
<summary>example</summary>

```
type ErrorWithCode interface {
   error
   ErrorCode() ErrorCode
}

type errorWithCode struct {
    errorCode ErrorCode
    err error
}

func (e errorWithCode) ErrorCode() ErrorCode {
  return e.errorCode
}
func (e errorWithCode) Error() string {
  return fmt.Sprintf("%d: %s", e.errorCode, e.err.Error())
}

func NewErrorCode(errorCode ErrorCode, err error) ErrorWithCode {
    return errorWithCode{errorCode: errorCode, err: err}
}
```


</details>

briandealwis commented on pkg/skaffold/errors/errors.go:
```diff
@@ -21,14 +21,15 @@ import (
+	Deploy      = Phase("Deploy")
+	StatusCheck = Phase("StatusCheck")
+	FileSync    = Phase("FileSync")
+	Dev         = Phase("Dev")
```
Is `Dev` really a phase?  (We don't actually use Dev from what I can tell?)

tejal29 replied:
it is not a phase, but dev-init is a stage or step in dev loop where errors can happen. 
I wanted to capture this "init" step errors.

Shd "DevInit" be better?

tejal29 replied:
also, i can rename "Phase" to "Stage"

tejal29 replied:
Change Dev -> DevInit
Add TearDown
Add comment saying these are phases in dev loop.

tejal29 commented on pkg/skaffold/event/event.go:
```diff
@@ -248,6 +248,27 @@ func BuildComplete(imageName string) {
+}
+
+// DevLoopFailed notifies that a dev loop has failed in a given phase
+func DevLoopFailedInPhase(i int, err error, phase sErrors.Phase) {
```
I try to follow the go naming conventions "Keep variable names short"
https://github.com/unknwon/go-code-convention/blob/master/en-US/naming_rules.md#variables

tejal29 replied:
changed the name

tejal29 commented on proto/skaffold.proto:
```diff
@@ -344,11 +353,32 @@ enum ErrorCode {
     STATUS_CHECK_NODE_NOT_READY= 406;
 
     // Unknown Error Codes
+    ErrorCode_UNKNOWN = 501;
```
`COULD_NOT_DETERMINE` or 0 value is default enum. In cases where we don't send any error Code the value is 0. To differentiate between when it is empty Vs when we know its an unknown error, i added the field.

tejal29 commented on proto/skaffold.proto:
```diff
@@ -344,11 +353,32 @@ enum ErrorCode {
     // Status Check error unknown
-    STATUS_CHECK_UNKNOWN = 501;
+    STATUS_CHECK_UNKNOWN = 502;
     // Container is unschedulable due to unknown reasons
-    STATUS_CHECK_UNKNOWN_UNSCHEDULABLE = 502;
+    STATUS_CHECK_UNKNOWN_UNSCHEDULABLE = 503;
     // Container is waiting due to unknown reason
-    STATUS_CHECK_CONTAINER_WAITING_UNKNOWN = 503;
+    STATUS_CHECK_CONTAINER_WAITING_UNKNOWN = 504;
```
so, I am trying to follow HTTP error codes (suggested by @balopat in one of the previous code review comments)
All unknowns are 5xx, 
All success codes are 2xx etc.

tejal29 commented:
Thanks a lot for all your feedback @briandealwis . I have addressed most of the comments and refactored method signatures. Defining a new struct `ActionableErr` makes sense so we skaffold stop (#3614) and dev iteration end event makes sense. 

I am wiring up Actionable Error messages here #4045 and will propagate suggestions in the events. 
Please take another look!

briandealwis commented on integration/rpc_test.go:
```diff
@@ -202,13 +205,16 @@ func TestEventLogHTTP(t *testing.T) {
 					deployEntries++
+				case *proto.Event_DevLoopEvent:
+					devLoopEntries++
 				default:
```
Please add a `t.Logf` here to we can catch possible flakes #4044 more quickly. (Or maybe a `t.Errorf` since they should be factored in?)

briandealwis commented on integration/rpc_test.go:
```diff
@@ -120,13 +120,16 @@ func TestEventsRPC(t *testing.T) {
 			deployEntries++
+		case *proto.Event_DevLoopEvent:
+			devLoopEntries++
 		default:
```
Please add a `t.Logf` here to we can catch possible flakes #4044 more quickly. (Or maybe a `t.Errorf` since they should be factored in?)

tejal29 replied:
done!

briandealwis commented on cmd/skaffold/app/cmd/cmd.go:
```diff
@@ -47,7 +47,7 @@ var (
 func NewSkaffoldCommand(out, err io.Writer) *cobra.Command {
 	updateMsg := make(chan string)
-	var shutdownAPIServer func() error
+	var 	shutdownAPIServer func() error
```
is this necessary?

tejal29 replied:
will remove

briandealwis commented on proto/skaffold.proto:
```diff
@@ -104,9 +112,33 @@ message Event {
+}
+
+// `ErrDef` defines an error occurred along with an optional suggestions
+message ActionableErr {
+    ErrorCode errCode = 1; // error code representing the error
+    string message = 2; // message describing the error.
+    repeated string suggestions = 3; // a list of suggestions for the error.
+
```
I don't think reporting messages here is the right thing to do.  I'm wondering if it would be better to have a secondary set of possible action codes that could be returned instead.  That would also allow i18n or more.

(I think it's also an opportunity to somehow segregate PII fields, like artifact names or image tags, so that we might be able to report the actual errors seen?)

tejal29 replied:
Thanks. I should have made this clear. The suggestion text is just for IDEs to show to User.
Having suggestions plumbed through skaffold will reduce the duplication of efforts.
I will add that as a comment.
For metrics, we should probably report codes like you mentioned.

briandealwis commented on proto/skaffold.proto:
```diff
@@ -310,7 +358,7 @@ enum ErrorCode {
     // No Error codes
```
I hit _Submit_ by accident.  One half-baked thought — which I think you're already considering — is that some of these ErrorCodes really identify the point of impact or result (e.g., the file-sync init error, or the dev-loop registeration errors), but sometimes not the actual error.

tejal29 replied:
I am going to add the `EndEvent` in another PR. Turns out this approach does not work with IDEs

tejal29 commented on pkg/skaffold/event/event.go:
```diff
@@ -380,6 +419,18 @@ func LogMetaEvent() {
 }
 
+// EndSessionEvent notifies that that skaffold session has ended and describes it
+func EndSessionEvent(errorCode proto.ErrorCode) {
```
This will be moved to another PR.

tejal29 commented on proto/skaffold.proto:
```diff
@@ -104,9 +112,33 @@ message Event {
     }
 }
 
+// `EndEvent` is the last event indicating skaffold session ended.
```
Will be moved to another PR and will address this there.

tejal29 commented on proto/skaffold.proto:
```diff
@@ -104,9 +112,33 @@ message Event {
+    ErrorCode errCode = 1; // error code representing the error
+    string message = 2; // message describing the error.
+    repeated string suggestions = 3; // a list of suggestions for the error.
+
```
remove this `suggestions`

tejal29 commented on pkg/skaffold/runner/dev.go:
```diff
@@ -27,14 +27,17 @@ import (
 
 // ErrorConfigurationChanged is a special error that's returned when the skaffold configuration was changed.
 var ErrorConfigurationChanged = errors.New("configuration changed")
+var iteration = 1
```
I can move this.

tejal29 commented on proto/skaffold.proto:
```diff
@@ -344,11 +392,32 @@ enum ErrorCode {
+    FILE_SYNC_INIT_ERROR = 601;
+
+
+    // Dev Monitor Errors
```
Change  `Dev Loop Initialization Errors`

tejal29 commented on pkg/skaffold/event/event.go:
```diff
@@ -493,6 +528,16 @@ func (ev *eventHandler) handle(event *proto.Event) {
+			logEntry.Entry = fmt.Sprintf("Dev Iteration %d in progress", de.Iteration)
+		case Succeeded:
+			logEntry.Entry = fmt.Sprintf("Dev Iteration %d successful", de.Iteration)
+		default:
```
change this `case Failed:`

tejal29 replied:
done

tejal29 commented on proto/skaffold.proto:
```diff
@@ -104,9 +104,18 @@ message Event {
 
+// `DevLoopEvent` marks the start and end of a dev loop.
+message DevLoopEvent {
+    int32 iteration = 1; // dev loop iteration
```
done!

briandealwis commented on pkg/skaffold/runner/dev.go:
```diff
@@ -105,22 +110,25 @@ func (r *SkaffoldRunner) doDev(ctx context.Context, out io.Writer) error {
 }
 
-// Dev watches for changes and runs the skaffold build and deploy
+// DevInit watches for changes and runs the skaffold build and deploy
```
oops this should still be Dev

briandealwis commented on proto/skaffold.proto:
```diff
@@ -104,9 +104,24 @@ message Event {
     }
 }
 
+// `DevLoopEvent` marks the start and end of a dev loop.
```
No backticks

briandealwis commented on proto/skaffold.proto:
```diff
@@ -104,9 +104,24 @@ message Event {
+
+}
+
+// `ErrDef` defines an error occurred along with an optional suggestions
```
No backticks

tejal29 replied:
All the protos struct defs have backticks in the description hence added these.

briandealwis commented on docs/content/en/api/skaffold.swagger.json:
```diff
@@ -380,6 +420,62 @@
+          },
+          {
+            "name": "event.devLoopEvent.err.errCode",
+            "description": " - UNKNOWN_ERROR: Could not determine error and phase\n - STATUSCHECK_SUCCESS: Status Check Success\n - STATUSCHECK_IMAGE_PULL_ERR: Container image pull error\n - STATUSCHECK_CONTAINER_CREATING: Container creating error\n - STATUSCHECK_RUN_CONTAINER_ERR: Container run error\n - STATUSCHECK_CONTAINER_TERMINATED: Container is already terminated\n - STATUSCHECK_CONTAINER_ ...
```
Can we do something to make this a more manageable string?

tejal29 replied:
this was recently changed in #4083 . I will follow up later.

briandealwis commented on proto/skaffold.proto:
```diff
@@ -302,53 +317,87 @@ enum ClusterType {
+    // FILE_SYNC Failures
+
+    // File Sync Initialize failure
+    SYNC_INIT_ERROR = 601;
```
This name seems slightly odd given the DEVINIT_ one below

tejal29 replied:
so, `Sync` represents the phase. and `INIT_ERROR` represents the error code. 
This error maps to initialization the sync file watcher. 
There are other possible errors in sync phase like failed to copy a file in container.
